hadoop-shims/hadoop2.8/src/main/java/org/apache/uniffle/hadoop/shim/HadoopShim28.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.uniffle.hadoop.shim;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapreduce.TaskAttemptID;
+import org.apache.hadoop.mapreduce.task.reduce.ShuffleClientMetrics;
+import org.apache.hadoop.mapreduce.v2.app.AppContext;
+import org.apache.hadoop.mapreduce.v2.app.client.ClientService;
+import org.apache.hadoop.mapreduce.v2.app.rm.RMContainerAllocator;
+import org.apache.hadoop.yarn.api.protocolrecords.AllocateResponse;
+import org.apache.hadoop.yarn.exceptions.YarnException;
+
+public class HadoopShim28 extends HadoopShim {
+
+  private static final Log LOG = LogFactory.getLog(HadoopShim28.class);
+
+  public ShuffleClientMetrics createShuffleClientMetrics(TaskAttemptID taskAttemptID, JobConf jobConf) {
+    try {
+      Constructor constructor =
```

Comment:
I can't get the point why we use reflection. In origin implement, we can use the constructor directly.

Reply:
Yes, I will change it. 
hadoop-shims is not introduced in my internal first version, I use reflection is that version. I forget to change it.

Reply:
The constructor of ShuffleClientMetrics has package-level access permissions, not public in hadoop-2.8.

Reply:
If we put this module to the mr module, we can avoid this problems. Because we use `org.apache.hadoop` in the mr module.